File: Day277_2025/Q277_GFG.py
class Solution:
    def help(self, s, target, index, expr, value, prev, ans):
        if index == len(s):
            if value == target:
                ans.append(expr)
            return

        num = 0
        for i in range(index, len(s)):
            num = num * 10 + int(s[i])
            temp = s[index:i+1]

            if index == 0:
                self.help(s, target, i + 1, temp, num, num, ans)
            else:
                self.help(s, target, i + 1, expr + "+" + temp, value + num, num, ans)
                self.help(s, target, i + 1, expr + "-" + temp, value - num, -num, ans)
                self.help(s, target, i + 1, expr + "*" + temp, value - prev + prev * num, prev * num, ans)

            # stop if number starts with 0
            if s[index] == "0":
                break

    def findExpr(self, s, target):
        ans = []
        self.help(s, target, 0, "", 0, 0, ans)
        return ans

File: Day277_2025/Q277_GFG.java
import java.util.ArrayList;

class Q277_GFG {
    private void help(String s, int target, int index, ArrayList<String> ans, long value, long prev, String expr) {
        if (index == s.length()) {
            if (value == target) {
                ans.add(expr);
            }
            return;
        }

        String temp = "";
        long num = 0;
        for (int i = index; i < s.length(); i++) {
            temp += s.charAt(i);
            num = num * 10 + (s.charAt(i) - '0');

            if (index == 0) {
                // first number, no operator before it
                help(s, target, i + 1, ans, num, num, temp);
            } else {
                help(s, target, i + 1, ans, value + num, num, expr + "+" + temp);
                help(s, target, i + 1, ans, value - num, -num, expr + "-" + temp);
                help(s, target, i + 1, ans, value - prev + prev * num, prev * num, expr + "*" + temp);
            }

            // avoid numbers with leading zeros
            if (s.charAt(index) == '0') break;
        }
    }

    public ArrayList<String> findExpr(String s, int target) {
        ArrayList<String> ans = new ArrayList<>();
        help(s, target, 0, ans, 0, 0, "");
        return ans;
    }
}
